packages/funding-rate-proposer/src/proposer.js

```diff
@@ -0,0 +1,360 @@
+      });
+      try {
+        const transactionResult = await runTransaction({
+          transaction: proposal,
+          config: {
+            gasPrice: this.gasEstimator.getCurrentFastPrice(),
+            from: this.account
+          }
+        });
+        let receipt = transactionResult.receipt;
+        let returnValue = transactionResult.returnValue.toString();
+
+        const logResult = {
+          tx: receipt.transactionHash,
+          requester: contractAddress,
+          proposer: this.account,
+          fundingRateIdentifier,
+          requestTimestamp,
+          proposedRate: offchainFundingRate,
+          currentRate: onchainFundingRate,
+          proposalBond: returnValue
+        };
+        this.logger.info({
+          at: "PerpetualProposer#updateFundingRate",
+          message: "Proposed new funding rate!🌻",
+          proposalBond: returnValue,
+          proposalResult: logResult
+        });
+      } catch (error) {
+        const message =
+          error.type === "call"
+            ? "Cannot propose funding rate: not enough collateral (or large enough approval)✋"
+            : "Failed to propose funding rate🚨";
+        this.logger.error({
+          at: "PerpetualProposer#updateFundingRate",
+          message,
+          fundingRateIdentifier,
+          error
+        });
+        return;
+      }
+    } else {
+      this.logger.debug({
+        at: "PerpetualProposer#updateFundingRate",
+        message: "Skipping proposal because current rate is within allowed margin of error",
+        fundingRateIdentifier,
+        proposedRate: offchainFundingRate,
+        currentRate: onchainFundingRate,
+        allowedError: this.fundingRateErrorPercent
+      });
+    }
+  }
+  // Sets allowances for all collateral currencies used live perpetual contracts.
+  async _setAllowances() {
+    // Get new approval receipts or null if approval was unneccessary.
+    const newApprovals = await Promise.map(Object.keys(this.contractCache), contractAddress => {
+      // The Perpetual requires approval to transfer the contract's collateral currency in order to post a bond.
+      // We'll set this once to the max value and top up whenever the bot's allowance drops below MAX_INT / 2.
+      return setAllowance(
+        this.web3,
+        this.gasEstimator,
+        this.account,
+        contractAddress,
+        this.contractCache[contractAddress].collateralAddress
+      );
+    });
+    newApprovals.forEach(receipt => {
+      if (receipt) {
+        this.logger.info({
+          at: "PerpetualProposer",
+          message: "Approved Perpetual contract to transfer unlimited collateral tokens 💰",
+          perpetual: receipt.spenderAddress,
+          currency: receipt.currencyAddress,
+          collateralApprovalTx: receipt.tx.transactionHash
+        });
+      }
+    });
+  }
+
+  async _cacheAndUpdatePriceFeeds() {
+    for (let contractAddress of Object.keys(this.contractCache)) {
```

Comment:
nit: thoughts on doing this as a big `Promise.map` so we can run these inner async calls in parallel?

Reply:
Nice catch definitely will do

Reply:
Bump.

Reply:
Done